fix: Re-issue certificate if key size changes

rjeffman opened the pull request:
**Resolves:** 

When requesting a certificate key size is not evaluated to consider a new certificate has to be requested.

**Result:**

This patch adds 'key_size' to the metadata comparison to determine if a new certificate request must be performed.

**Issue Tracker Tickets (Jira or BZ if any):** [RHBZ#2186057](https://bugzilla.redhat.com/show_bug.cgi?id=2186057)

rjeffman commented:
[citest]

richm commented on tests/tests_key_size_reissue.yml:
```diff
@@ -0,0 +1,64 @@
+      ansible.builtin.import_role:
+        name: linux-system-roles.certificate
+      register: result
+      failed_when: not result.changed
```
I don't think this works - that is, `import_role` and `include_role` do not report `changed` if the last task reported `changed` - if you really want to test it this way, you'll need to register the result of `certificate_request` to a global role variable that you can export - we do something similar in the firewall role - https://github.com/linux-system-roles/firewall/blob/main/tasks/main.yml#L104 - then use it like https://github.com/linux-system-roles/firewall/blob/2e43f8a1e240d0b89775a61776d2377aa1c91113/tests/tests_zone.yml#L30


* add a `register: __certificate_request_result` to `name: Ensure certificate requests` - I think variables that are registered go into the public namespace, so you should be able to use it from the test
* in the tests, call the role with `include_role` (not `import_role`) and add `public: true` to export the privat

richm replied:
hmm - well, the rhel 8.9 test passed - but I'm not even sure if the `failed_when` is being evaluated

rjeffman replied:
yes, I forgot to remove the `register` and `failed_when` directives before opening the PR.

I'll fix it along with the CentOS/RHEL 7 fix.

rjeffman replied:
fixed

rjeffman commented:
I fixed the issues with CentOS/RHEL 7, and rewrote the tests for `key_size` as they were not properly testing scenarios where certmonger version was less than `0.79.0`.

rjeffman commented:
[citest]

rjeffman commented:
[citest]

richm commented on library/certificate_request.py:
```diff
@@ -359,7 +358,7 @@ def _get_argument_spec():
             provider_config_directory=dict(type="str", default="/etc/certmonger"),
             provider=dict(type="str", default="certmonger"),
-            key_size=dict(type="int", default=2048),
+            key_size=dict(type="int", default=None),
```
so the certmonger provider will by default issue 2048 key_size?  On el7, el8, el9, and fedora?

rjeffman replied:
Yes, by default it will be 2048.

Also, there's a warning in the role documentation that if the user wants to ensure a specific key size, the parameter `key_size` should be explicitly used as the default value could be changed depending on the role version.
